alf/algorithms/algorithm.py

```diff
@@ -78,13 +115,111 @@ def __init__(self,
         if predict_state_spec is None:
             predict_state_spec = train_state_spec
         self._predict_state_spec = predict_state_spec
-        self._optimizer = optimizer
+        self._is_rnn = len(tf.nest.flatten(train_state_spec)) > 0
+
+        if not optimizer:
+            assert trainable_module_sets is None
+            self._init_optimizers = [None]
+            self._init_module_sets = [[]]
+        elif isinstance(optimizer, tf.optimizers.Optimizer):
+            assert trainable_module_sets is None
+            self._init_optimizers = [optimizer]
+            self._init_module_sets = [[]]
+        else:
+            assert isinstance(optimizer, list)
+            assert isinstance(trainable_module_sets, list)
+            assert len(trainable_module_sets) == len(optimizer), (
+                "`optimizer` and `trainable_module_sets`"
+                "should have same length")
+            self._init_optimizers = optimizer
+            self._init_module_sets = trainable_module_sets
+
+        self._cached_opt_and_var_sets = None
+        self._gradient_clipping = gradient_clipping
+        self._clip_by_global_norm = clip_by_global_norm
         self._debug_summaries = debug_summaries
 
     @property
-    def optimizer(self):
-        """Return the optimizer for this algorithm."""
-        return self._optimizer
+    def trainable_variables(self):
+        return sum([var_set for _, var_set in self._get_opt_and_var_sets()],
```

Comment:
`_get_opt_and_var_sets` can be replaced with  `_get_cached_opt_and_var_sets`

Reply:
It is intentionally this to have the same behavior as tf.Module.trainable_variables. It is expected to always give the correct result even with after new submodules are added.